src/gax/src/error/core_error.rs

```diff
@@ -12,161 +12,664 @@
+/// Applications can query the error [source][std::error::Error::source] for
+/// deeper information.
+///
+/// # Example
+/// ```
+/// use google_cloud_gax::error::Error;
+/// match example_function() {
+///     Err(e) if e.is_service() => { println!("service error {e}, should have a status={:?}", e.status()); },
+///     Err(e) => { println!("some other error {e}"); },
+///     Ok(_) => { println!("success, how boring"); },
+/// }
+///
+/// fn example_function() -> Result<String, Error> {
+///     // ... details omitted ...
+///     # use google_cloud_gax::error::rpc::{Code, Status};
+///     # Err(Error::service(Status::default().set_code(Code::NotFound).set_message("NOT FOUND")))
+/// }
+/// ```
 #[derive(Debug)]
 pub struct Error {
     kind: ErrorKind,
-    source: BoxError,
 }
 
 impl Error {
-    /// Creates a new [Error] with the given [ErrorKind] and source error.
-    pub fn new<T: Into<BoxError>>(kind: ErrorKind, source: T) -> Self {
-        Error {
-            kind,
-            source: source.into(),
+    /// Creates an error with the information returned by Google Cloud services.
+    ///
+    /// # Example
+    /// ```
+    /// use google_cloud_gax::error::Error;
+    /// use google_cloud_gax::error::rpc::{Code, Status};
+    /// let status = Status::default().set_code(Code::NotFound).set_message("NOT FOUND");
+    /// let error = Error::service(status.clone());
+    /// assert!(error.is_service());
+    /// assert_eq!(error.status(), Some(&status));
+    /// ```
+    pub fn service(status: Status) -> Self {
+        let kind = ErrorKind::Service {
+            status,
+            status_code: None,
+            headers: None,
+        };
+        Self { kind }
+    }
+
+    /// The error was returned by the service.
+    pub fn is_service(&self) -> bool {
+        matches!(&self.kind, ErrorKind::Service { .. })
+    }
+
+    /// The [Status] payload associated with this error.
+    ///
+    /// # Examples
+    /// ```
+    /// use google_cloud_gax::error::{Error, rpc::{Code, Status}};
+    /// let error = Error::service(Status::default().set_code(Code::NotFound));
+    /// if let Some(status) = error.status() {
+    ///     if status.code == Code::NotFound {
+    ///         println!("cannot find the thing, more details in {:?}", status.details);
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// Google Cloud services return a detailed `Status` message including a
+    /// numeric code for the error type, a human-readable message, and a
+    /// sequence of details which may include localization messages, or more
+    /// information about what caused the failure.
+    ///
+    /// See [AIP-193] for background information about the error model in Google
+    /// Cloud services.
+    ///
+    /// [AIP-193]: https://google.aip.dev/193
+    pub fn status(&self) -> Option<&Status> {
+        match &self.kind {
+            ErrorKind::Service { status, .. } => Some(status),
+            _ => None,
```

Comment:
Q: Will we translate random HTTP status codes to a `Status` code?

I feel like we shouldn't, for the sake of predictability. Then you know if `status() == None`, the request likely did not make it to the service.

Reply:
> Q: Will we translate random HTTP status codes to a Status code?

No.

> I feel like we shouldn't, for the sake of predictability.

Agree, though that will take some more work to fix gRPC.

> Then you know if status() == None, the request likely did not make it to the service.

"Likely" maybe, but you cannot assert that it did not. If you can, you can solve the two general's problem.